byron/ledger/executable-spec/test/Ledger/Relation/Properties.hs

```diff
@@ -0,0 +1,223 @@
+-- Properties on Relations
+--------------------------------------------------------------------------------
+
+-- | (dom r) ∩ s == dom (s ◁ r)
+--
+--  By restricting the domain of 'Relation r' to 's', the new 'Domain r'
+--  is an intersection of the original 'Domain r' with 's'
+propDomainRestrictionAndIntersection
+  :: (MonadTest m, Relation r, Ord (Domain r))
+  => Domain r -> Set (Domain r) -> r -> m ()
+propDomainRestrictionAndIntersection x s r
+  = Set.member x (dom r `Set.intersection` s)
+    ===
+    Set.member x (dom (s ◁ r))
+
+-- | (dom r ∩ s) ◁ r == s ◁ r
+--
+--  Restricting the domain of 'Relation r' to 's'
+--  is the same as restricting  the domain to 'dom r ∩ s'.
+propDomainRestrictionAndIntersectionB
+  :: (MonadTest m, Relation r, Ord (Domain r), Eq (Range r), Show (Range r))
+  => (Domain r -> r -> Maybe (Range r))
+  -> Domain r
+  -> Set (Domain r)
+  -> r
+  -> m ()
+propDomainRestrictionAndIntersectionB find x s r
+  = find x ((dom r `Set.intersection` s) ◁ r)
+    ===
+    find x (s ◁ r)
+
+-- | (dom r \\ s) ◁ r == s ⋪ r
+--
+--  Excluding a from the domain of 'Relation r'
+--  is the same as restricting the domain to 'dom r \\ s'.
+propDomainExclusionAndSetDifference
+  :: (MonadTest m, Relation r, Ord (Domain r), Eq (Range r), Show (Range r))
+  => (Domain r -> r -> Maybe (Range r))
+  -> Domain r
+  -> Set (Domain r)
+  -> r
+  -> m ()
+propDomainExclusionAndSetDifference find x s r
+  = find x ((dom r `Set.difference` s) ◁ r)
+    ===
+    find x (s ⋪ r)
+
+-- | (dom r1 ∪ s) ⋪ (r1 ∪ r2) == (dom r1 ∪ s) ⋪ r2
+--
+--  Excluding 'dom r1 ∪ s' from the domain of 'r1 ∪ r2'
+--  is the same as excluding 'dom r1 ∪ s' from the domain of just r2
+--  (since 'dom r1 ⋪ r1' gives an empty relation)
+propDomainExclusionAndUnion
+  :: (MonadTest m, Relation r, Ord (Domain r), Ord (Range r), Show (Range r))
+  => (Domain r -> r -> Maybe (Range r))
+  -> Domain r
+  -> Set (Domain r)
+  -> r
+  -> r
+  -> m ()
+propDomainExclusionAndUnion find x s r1 r2
+  = find x ((dom r1 `Set.union` s) ⋪ (r1 ∪ r2))
+    ===
+    find x ((dom r1 `Set.union` s) ⋪ r2)
+
+
+-- | s ◁ (r1 ∪ r2) == (s ◁ r1) ∪ (s ◁ r2)
+--
+-- Restricting the domain of 'r1 ∪ r2' to 's' is the same as restricting the
+-- domains of r1 and r2 seperately and then taking the relation union of that.
+{- TODO @uroboros keep this?
+propDomainRestrictionAndUnion
+  :: (MonadTest m, Relation r, Ord (Domain r), Ord (Range r), Show (Range r))
+  => (Domain r -> r -> Maybe (Range r))
+  -> Domain r
+  -> Set (Domain r)
+  -> r
+  -> r
+  -> m ()
+propDomainRestrictionAndUnion find x s r1 r2
+  = find x (s ◁ (r1 ∪ r2))
+    ===
+    find x ((s ◁ r1) ∪ (s ◁ r2))
+-}
+
+--------------------------------------------------------------------------------
+-- Functions abstracted over 'Relation r'
+--------------------------------------------------------------------------------
+
+-- | Combined property abstracted over 'Relation r'
```

Comment:
I'm afraid what we're introducing a lot of dependencies between the generators by using monadic binds. Couldn't we test the properties separately?

Reply:
👍